File: java/src/com/ctre/phoenix/ButtonMonitor.java
package com.ctre.phoenix;

import edu.wpi.first.wpilibj.GenericHID;

public class ButtonMonitor implements ILoopable
{
	GenericHID  _gameCntrlr;
	int _btnIdx;
	IButtonPressEventHandler  _handler;
	boolean _isDown = false;
	
	public interface IButtonPressEventHandler {
		void OnButtonPress(int idx, boolean isDown);
	};

	public ButtonMonitor(GenericHID controller, int buttonIndex,
			IButtonPressEventHandler ButtonPressEventHandler) {
		_gameCntrlr = controller;
		_btnIdx = buttonIndex;
		_handler = ButtonPressEventHandler;
	}

	public void Process() {
		boolean down = _gameCntrlr.getRawButton(_btnIdx);

		if (!_isDown && down){
			_handler.OnButtonPress(_btnIdx, down);
		}

		_isDown = down;
	}

	public void OnStart() {
	}
	public void OnLoop() {
		Process();
	}
	public boolean IsDone() {
		return false;
	}
	public void OnStop() {
	}
}

File: cpp/src/Tasking/ButtonMonitor.cpp
#include "ctre/phoenix/Tasking/ButtonMonitor.h"

#ifndef CTR_EXCLUDE_WPILIB_CLASSES
#include <GenericHID.h> // WPILIB

namespace ctre {
namespace phoenix {
namespace tasking {

ButtonMonitor::ButtonMonitor(frc::GenericHID * controller, int buttonIndex,
		IButtonPressEventHandler * ButtonPressEventHandler) {
	_gameCntrlr = controller;
	_btnIdx = buttonIndex;
	_handler = ButtonPressEventHandler;
}
ButtonMonitor::ButtonMonitor(const ButtonMonitor & rhs) {
	_gameCntrlr = rhs._gameCntrlr;
	_btnIdx = rhs._btnIdx;
	_handler = rhs._handler;
}

void ButtonMonitor::Process() {
	bool down = ((frc::GenericHID*)_gameCntrlr)->GetRawButton(_btnIdx);

	if (!_isDown && down)
		_handler->OnButtonPress(_btnIdx, down);

	_isDown = down;
}

void ButtonMonitor::OnStart() {
}
void ButtonMonitor::OnLoop() {
	Process();
}
bool ButtonMonitor::IsDone() {
	return false;
}
void ButtonMonitor::OnStop() {
}

} // namespace tasking
} // namespace phoenix
} // namespace ctre

#endif // CTR_EXCLUDE_WPILIB_CLASSES
